### PR TITLE
Basic 2d functionality

### PR DESCRIPTION
**Pull request by Bencargs**
### Demonstrated in this PR:
Initial commit showing basic map editing functionality over 2d game world.

- Tile based world representation
- Tile selection via spatial hashing
- Entity-Component architecture
- Physics simulated movement
- Collision detection using both BoundingCircle and AABB
- Image sprite and image animation
- Hierarchical finite state machine - removed for units in favour of EC design, could be reused for game state.
- Type Object representation on unit and map data

image

## To Do:
- Many areas to refactor and improve

Main areas of focus in next PR
- Map / unit saving & loading
- Unit controller
- Camera handling
- Path finding
- Particle System
- Picking & Selection
- Command Handler
- Event Messaging System
- Fog of War / LOS system
- Replace Forms with Game class
- Replace WinForms with XNA
- Convert to 3d

### CODE REVIEW

File: MapEditor/Components/CollisionController.cs [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+            // foreach cell from vector start to vector end
+            // get collider - check collission - if true return, else keep going
+            // return null
+            var tiles = _map.GetTiles(line);
+            foreach (var t in tiles)
+            {
+                //line circle intersection - https://yal.cc/gamemaker-collision-line-point/
+                //https://stackoverflow.com/questions/23016676/line-segment-and-circle-intersection
+                //return t.Colliders.FirstOrDefault();
+                foreach (var c in t.Colliders)
```
Comment (Bencargs) [inline MapEditor/Components/CollisionController.cs]: Either implement linked example, or uncomment above line,

File: MapEditor/Common/Vector2.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
+﻿namespace MapEditor.Common
+{
+    // todo: temporary - replace when using a real framework
+    public class Vector2
```
Comment (Bencargs) [inline MapEditor/Common/Vector2.cs]: Temporary stub class

File: MapEditor/Components/CollisionController.cs [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        private bool IsCollided(BoundingCircle collider)
+        {
+            var distanceX = Position.X - collider.Position.X;
+            var distanceY = Position.Y - collider.Position.Y;
+
+            var magnitudeSquared = distanceX * distanceX + distanceY * distanceY;
+            return magnitudeSquared < (Radius + collider.Radius) * (Radius + collider.Radius);
+        }
+
+        private bool IsCollided(BoundingBox box)
```
Comment (Bencargs) [inline MapEditor/Components/CollisionController.cs]: Think about how these three situations can be make more generic for code sharing

File: MapEditor/Components/MovementHandler.cs [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+            var velocity = movementData.Velocity;
+            var slowRadius = movementData.SlowRadius;
+            var stopRadius = movementData.StopRadius;
+            var maxVelocity = movementData.MaxVelocity;
+            var force = movementData.Force;
+            var mass = physicsData.Mass;
+
+            ITarget target;
+            Vector2 steering = null;    //todo: save previous steering, then do steering += Move()
+            switch (movementData.MovementMode)
```
Comment (Bencargs) [inline MapEditor/Components/MovementHandler.cs]: Very long class - should move switch into seperate method? of make MovementData a class and have it implement its own movement?

File: MapEditor/Components/MovementHandler.cs [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+            return Move(targetVector,
+                position,
+                velocity,
+                slowRadius,
+                maxVelocity,
+                force,
+                mass);
+        }
+
+        private void Follow(ITarget target)
```
Comment (Bencargs) [inline MapEditor/Components/MovementHandler.cs]: There is examples online for this, will be important for group movement.
Leader + Seperation

File: MapEditor/Engine/Map.cs [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+        }
+
+        // todo: replace bresenhams with supercover algorithm
+        public IEnumerable<Tile> GetTiles(Vector2 position)
+        {
+            // https://stackoverflow.com/questions/11678693/all-cases-covered-bresenhams-line-algorithm
+            var startPoint = new Point(position.X, position.Y);
+            var endPoint = startPoint.Add(position.Normalize() * position.Length());
+
+            var w = endPoint.X - startPoint.X;
```
Comment (Bencargs) [inline MapEditor/Engine/Map.cs]: Move this to a common helper class?

File: MapEditor/Engine/Map.cs [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+                }
+                else
+                {
+                    x += dx2;
+                    y += dy2;
+                }
+            }
+        }
+
+        //// todo: compare against http://ericw.ca/notes/bresenhams-line-algorithm-in-csharp.html
```
Comment (Bencargs) [inline MapEditor/Engine/Map.cs]: These three methods require testing

File: MapEditor/Entities/ComponentController.cs [hunk 8]
```diff
@@ -0,0 +1,118 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Drawing;
+using MapEditor.Common;
+using MapEditor.Components;
+
+namespace MapEditor.Entities
+{
+    public interface IComponent
```
Comment (Bencargs) [inline MapEditor/Entities/ComponentController.cs]: Clean up this class

File: MapEditor/Entities/UnitController.cs [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using MapEditor.Engine;
+
+namespace MapEditor.Entities
+{
+    public class UnitController
+    {
+        private readonly List<Entity> _units;
```
Comment (Bencargs) [inline MapEditor/Entities/UnitController.cs]: Ensure units are properly object pooled.
track last active unit flag

File: MapEditor/Form1.cs [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+
+        private void Button1_Click(object sender, EventArgs e)
+        {
+            var dialog = new OpenFileDialog
+            {
+                Filter = @"Image Files|*.bmp",
+                Title = @"Select a Background File"
+            };
+
+            if (dialog.ShowDialog() == DialogResult.OK)
```
Comment (Bencargs) [inline MapEditor/Form1.cs]: Complete this

File: MapEditor/Form1.cs [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+                thumbnail.Location = canvas.PointToClient(Cursor.Position);
+                thumbnail.Image = terrain.Image;
+                thumbnail.BringToFront();
+                thumbnail.Visible = true;
+                return;
+            }
+            e.Effect = DragDropEffects.None;
+        }
+
+        private void Button2_Click(object sender, EventArgs e)
```
Comment (Bencargs) [inline MapEditor/Form1.cs]: Rename

File: MapEditor/JsonImageConverter.cs [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+using System.Drawing;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+
+namespace MapEditor
+{
+    public class JsonImageConverter : JsonConverter
```
Comment (Bencargs) [inline MapEditor/JsonImageConverter.cs]: Unused - delete

File: MapEditor/TileForm.cs [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+
+            _terrain = terrain;
+            var comboItems = ((TerrainType[]) Enum.GetValues(typeof(TerrainType))).Select(x => new FormattedTerrainItem(x));
+            terrainCmb.Items.AddRange(comboItems.ToArray());
+            terrainCmb.SelectedIndex = (int)_terrain.TerrainType;
+            terrainCmb.ValueMember = "Value";
+            terrainCmb.DisplayMember = "Description";
+        }
+
+        private void ComboBox1_SelectedIndexChanged(object sender, EventArgs e)
```
Comment (Bencargs) [inline MapEditor/TileForm.cs]: ?

File: MapEditor/TileForm.cs [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+                //todo: _map.SetTerrain
+                var previous = _terrain;
+                var type = ((FormattedTerrainItem) terrainCmb.SelectedItem).Value;
+                _terrain = new Terrain(type, previous.Image, previous.Width, previous.Height);
+            }
+            Close();
+        }
+    }
+
+    internal sealed class FormattedTerrainItem
```
Comment (Bencargs) [inline MapEditor/TileForm.cs]: This should be a temporary concern - move to an extension class?

**Comment by BenCargill** [comment]
Other required modules:
- Audio
- Text & Font controller, Console Input and rendering
- Trigger / Event system
- AI Sytem
- Finite State machine
- Game State?
- UI manager
- Heightmap
- Effects system
- Database?

File: MapEditor/Components/CollisionController.cs (hunk 3 above)
Comment (Bencargs) [reply MapEditor/Components/CollisionController.cs]: Done via extension method

**Pull request merged**